graphql_compiler/macros/__init__.py

```diff
@@ -1 +1,411 @@
+    reverse_target_class_name = macro_descriptor.base_class_name
+
+    existing_descriptor = _find_macro_edge_name_at_subclass(
+        macro_registry, reverse_base_class_name, reverse_macro_edge_name)
+
+    if existing_descriptor is not None:
+        # There is already a reverse macro edge of the same name that starts at the same type.
+        # Let's make sure its endpoint types are an exact match compared to the endpoint types
+        # of the macro edge being defined.
+        if reverse_base_class_name != existing_descriptor.base_class_name:
+            raise GraphQLInvalidMacroError()
+
+        if reverse_target_class_name != existing_descriptor.target_class_name:
+            raise GraphQLInvalidMacroError()
+
+    existing_descriptor = _find_macro_edge_name_to_subclass(
+        macro_registry, reverse_target_class_name, reverse_macro_edge_name)
+    if existing_descriptor is not None:
+        # There is already a macro edge of the same name that points to the same type.
+        # Let's make sure its endpoint types are an exact match compared to the endpoint types
+        # of the macro edge being defined.
+        if reverse_target_class_name != existing_descriptor.target_class_name:
+            raise GraphQLInvalidMacroError()
+
+        if reverse_base_class_name != existing_descriptor.base_class_name:
+            raise GraphQLInvalidMacroError()
+
+
+def register_macro_edge(macro_registry, macro_edge_graphql, macro_edge_args):
+    """Add the new macro edge descriptor to the provided MacroRegistry object, mutating it.
+
+    Args:
+        macro_registry: MacroRegistry object containing macro descriptors, where the new
+                        macro edge descriptor should be added.
+        macro_edge_graphql: string, GraphQL defining how the new macro edge should be expanded
+        macro_edge_args: dict mapping strings to any type, containing any arguments the macro edge
+                         requires in order to function.
+    """
+    # The below function will validate that the macro edge in question is valid in isolation,
+    # when considered only against the macro-less schema. After geting this result,
+    # we simply need to check the macro edge descriptor against other artifacts in the macro system
+    # that might also cause conflicts.
+    macro_descriptor = make_macro_edge_descriptor(
+        macro_registry.schema_without_macros, macro_registry.subclass_sets,
+        macro_edge_graphql, macro_edge_args,
+        type_equivalence_hints=macro_registry.type_equivalence_hints)
+
+    # Ensure there's no conflict with macro edges defined on subclasses and superclasses.
+    _check_macro_edge_for_definition_conflicts(macro_registry, macro_descriptor)
+
+    # Ensure there's no conflict between existing macro edges and the (hypothetical) reversed
+    # macro edge of the one being defined.
+    _check_macro_edge_for_reversal_definition_conflicts(macro_registry, macro_descriptor)
```

Comment:
Maybe returning all conflicts is more user friendly. Imagine fixing a naming conflict only to find out there's another naming conflict. I won't block for this though, it's definitely a "nice to have".

Reply:
I agree, it's a nice to have. In the interest of getting the rest of this out, though, I think I'll save it for another time – and hopefully having multiple conflicts doesn't happen often enough in anyone's workflow for this to really matter in the end.